app/api/currentuser/export_data.go

```diff
@@ -0,0 +1,200 @@
+package currentuser
+
+import (
+	"encoding/json"
+	"io"
+	"net/http"
+	"strconv"
+	"strings"
+
+	"github.com/France-ioi/AlgoreaBackend/app/database"
+	"github.com/France-ioi/AlgoreaBackend/app/service"
+)
+
+// swagger:operation GET /current-user/export-data users currentUserDataExport
+// ---
+// summary: Export the current user's data
+// description: >
+//   Returns a downloadable JSON file with all the current user's data.
+//   The content returned is just the dump of raw entries of each table related to the user
+//
+//     * `current_user` (from `users`): all attributes except `iVersion`
+//     * `sessions`, `refresh_token`: all attributes, but secrets replaced with “***”
+//     * `owned_groups`: `ID` and `sName` for every descendant of user’s `idGroupOwned`;
+//     * `joined_groups`: `ID` and `sName` for every ancestor of user’s `idGroupSelf`;
+//     * `users_answers`: all attributes;
+//     * `users_items`: all attributes except `iVersion`;
+//     * `groups_attempts`: the user's or his teams' attempts, all attributes except `iVersion`;
+//     * `groups_groups`: where the user’s `idGroupSelf` is the `idGroupChild`, all attributes except `iVersion` + `groups.sName`.
+// produces:
+//   - application/json
+// responses:
+//   "200":
+//     description: The returned data dump file
+//     schema:
+//       type: file
+//   "401":
+//     "$ref": "#/responses/unauthorizedResponse"
+//   "500":
+//     "$ref": "#/responses/internalErrorResponse"
+func (srv *Service) exportData(w http.ResponseWriter, r *http.Request) service.APIError {
+	user := srv.GetUser(r)
+
+	w.Header().Set("Content-Type", "application/json; charset=utf-8")
+	w.Header().Set("Content-Disposition", "attachment; filename=user_data.json")
+	w.WriteHeader(200)
+
+	databaseName := srv.Config.Database.Connection.DBName
+
+	_, err := w.Write([]byte("{"))
```

Comment:
Here we cannot generate the whole response at once (as we do in other services) because it the result can be pretty large (tens of megabytes) and much more (gigabytes) memory is needed to process all the data at once. So we need to stream the data without keeping it in memory.

Reply:
I get it. Great 👍